activerecord/test/cases/collection_cache_key_test.rb

```diff
@@ -0,0 +1,62 @@
+require "cases/helper"
+require "models/computer"
+require "models/developer"
+require "models/project"
+require "models/topic"
+require "models/post"
+require "models/comment"
+
+module ActiveRecord
+  class CollectionCacheKeyTest < ActiveRecord::TestCase
+    fixtures :developers,  :projects, :developers_projects, :topics, :comments, :posts
+
+    test "collection_cache_key on model" do
+      assert_match(/\Adevelopers\/query-(\h+)-(\d+)-(\d+)\Z/, Developer.collection_cache_key)
+    end
+
+    test "collection_cache_key with custom collection" do
+      collection = Topic.where(author_name: "Carl")
+      assert_match(/\Atopics\/query-(\h+)-(\d+)-(\d+)\Z/, ActiveRecord::Base.collection_cache_key(collection))
+    end
+
+    test "cache_key for relation" do
+      developers = Developer.where(name: "David")
+      assert_match(/\Adevelopers\/query-(\h+)-(\d+)-(\d+)\Z/, developers.cache_key)
+    end
+
+    test "it triggers at most one query" do
+      developers =  Developer.where(name: "David")
+
+      assert_queries(1) { developers.cache_key }
+      assert_queries(0) { developers.cache_key }
+    end
+
+    test "it doesn't trigger any query if the relation is already loaded" do
+      developers =  Developer.where(name: "David").load
+      assert_queries(0) { developers.cache_key }
+    end
+
+    test "relation cache_key changes when the sql query changes" do
+      developers = Developer.where(name: "David")
+      other_relation =  Developer.where(name: "David").where("1 = 1")
+
+      assert_not_equal developers.cache_key, other_relation.cache_key
+    end
+
+    test "cache_key for empty relation" do
+      developers = Developer.where(name: "Non Existent Developer")
+      assert_match(/\Adevelopers\/query-(\h+)-0\Z/, developers.cache_key)
+    end
```

Comment:
[I've added these](https://github.com/afcapel/rails/blob/e1120a74d164c51568c18fdb8bc5d6471b1a1d2b/activerecord/test/cases/collection_cache_key_test.rb#L28-L32), to check each part of the cache key. Is that ok?